### PR TITLE
Database

### PR DESCRIPTION
**Pull request by Nikitak211**

### CODE REVIEW

File: api/registerRouter.js [hunk 1]
```diff
@@ -4,7 +4,7 @@ const router = express.Router();
 // const mydata = require('../models/mydata');
 
 router.post('/register', (req, res) => {
-    res.sen(console.log('hello'))
+    res.send(console.log('hello'))
```
Comment (8o8inCodes) [inline api/registerRouter.js]: Don't send console.log, what you're actually sending is whatever console.log returns.
Send instead just a string "hello".

File: app.js [hunk 2]
```diff
@@ -2,24 +2,121 @@ const express = require('express');
 const path = require('path');
 const register = require('./api/registerRouter');
 const app = express();
+const mongoose = require('mongoose');
+const AuthRoute = require('./routes/auth')
+const session = require('express-session');
+const User = require('./models/user')
+const bcrypt = require('Bcryptjs')
+const jwt = require('jsonwebtoken')
+const MongoDBStore = require('connect-mongodb-session')(session);
```
Comment (8o8inCodes) [inline app.js]: Sort out the imports:
Third party packages first, your imports second

File: app.js [hunk 3]
```diff
@@ -2,24 +2,121 @@ const express = require('express');
...
+mongoose.Promise = global.Promise
 
-//body parser Midware
-app.use(express.json());
-app.use(express.urlencoded({ extended: false }));
+const mongoDBUri = 'mongodb+srv://PosterApp:A1s2f4g5%2A@poster-data.noemv.mongodb.net/Data'
```
Comment (8o8inCodes) [inline app.js]: Bad, you saved a username and password in git (it's gonna be here forever now, now everyone can see it. You need to save it in environment variable.
https://www.npmjs.com/package/dotenv
Read about this ^
And make sure you add that .env file to the gitignore. At the same time create an example.env file with all the keys that are used by the project (without values)

File: app.js [hunk 4]
```diff
@@ -2,24 +2,121 @@ const express = require('express');
...
+}
+
+app.get('/homepage',isAuth , (req, res) => {
     res.sendFile(path.resolve(__dirname,'public/homepage/homepage.html'))
 })
 
-app.get('/login', (req, res) => {
+app.get('/', (req, res) => {
     res.sendFile(path.resolve(__dirname,'public/login/login.html'))
 })
```
Comment (8o8inCodes) [inline app.js]: You don't need the /homepage route
in "/" you need to check if user logged in or not, and sendFile depending if the user is logged in or not.

File: app.js [hunk 5]
```diff
@@ -2,24 +2,121 @@ const express = require('express');
...
+
+    const isMatch = await bcrypt.compare(password, user.password)
+
+    if (!isMatch) {
+        return  res.redirect('/');
+    }
+
+    req.session.isAuth = true
+    res.redirect('/homepage');
+})
```
Comment (8o8inCodes) [inline app.js]: I am assuming that this is a login api, it has to be in the api routes file. not in app.js.

Also since this is an api, res.redirect is not gonna work, because the script is being redirected instead of the browser. You will need to handle it differently, by sending an object {success: true} or something and then in frontend check if it succeeded then redirect to "/" (homepage)

File: app.js [hunk 6]
```diff
@@ -2,24 +2,121 @@ const express = require('express');
...
+    user = new User({
+        username,
+        password: hashedPass,
+        email,
+        lastName,
+        firstName
+    })
+    await user.save();
+    res.redirect('/')
+});
```
Comment (8o8inCodes) [inline app.js]: Should be in api routes, and same thing with the res.redirect

File: app.js [hunk 7]
```diff
@@ -2,24 +2,121 @@ const express = require('express');
...
+
+
+app.post('/logout', (req, res) => {
+    req.session.destroy((err) => {
+        if (err) throw err;
+        res.redirect('/');
+    })
+})
```
Comment (8o8inCodes) [inline app.js]: apiroutes

File: app.js [hunk 8]
```diff
@@ -2,24 +2,121 @@ const express = require('express');
...
 const port = process.env.PORT || 5000;
-app.listen(port, () => console.log(`Listening on port ${port}...`));
\ No newline at end of file
+app.listen(port, () => console.log(`Listening on port ${port}...`));
+
+
+//body parser Midware
+app.use(express.json());
+app.use(express.urlencoded({ extended: false }));
+app.use('/api',AuthRoute)
```
Comment (8o8inCodes) [inline app.js]: Rename authroute to ApiRoutes

File: controller/AuthControll.js [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+            res.json({
+                message: "User Added to database Successfully"
+            })
+        })
+    })
+}
+
+const login = (req, res, next) => {
+    var username = req.body.username
+    var password = req.body.password
```
Comment (8o8inCodes) [inline controller/AuthControll.js]: Don't use vars, use either const or let

File: controller/AuthControll.js [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+    User.findOne({$or: [
+        {email:username},
+        {firstName:username},
+        {username:username},
+        {lastName:username}
```
Comment (8o8inCodes) [inline controller/AuthControll.js]: What is this? you need to find only username: username, without $or

File: controller/AuthControll.js [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    .then( user => {
+        if ( user ) {
+            bcrypt.compare(password, user.password, function(err, result) {
+                if (err) {
+                    res.json({
+                       error: err 
+                    })
+                }
+                if ( result ) {
+                    let token = jwt.sign({name: user.name}, ')"ĩĐ(ĐõÛ}vß}ĩ', {expiresIn: '1h'})
```
Comment (8o8inCodes) [inline controller/AuthControll.js]: secret should be in the .env as well

File: middleware/authenticate.js [hunk 12]
```diff
@@ -0,0 +1,17 @@
+const jwt = require('jsonwebtoken')
+
+const authenticate = (req, res, next) => {
+    try {
+        const token = req.headers.authorization.split(' ')[1]
+        const decode = jwt.verify(token, ')"ĩĐ(ĐõÛ}vß}ĩ')
```
Comment (8o8inCodes) [inline middleware/authenticate.js]: move secret to .env

File: package.json [hunk 13]
```diff
@@ -17,7 +17,12 @@
   },
   "homepage": "https://github.com/Nikitak211/poster#readme",
   "dependencies": {
+    "bcryptjs": "^2.4.3",
+    "connect-mongodb-session": "^3.1.0",
     "express": "^4.17.1",
+    "express-session": "^1.17.2",
+    "jsonwebtoken": "^8.5.1",
+    "mocha": "^9.1.2",
```
Comment (8o8inCodes) [inline package.json]: Why are you using mocha? You don't write unit tests yet.

File: public/login/login.html [hunk 14]
```diff
@@ -18,7 +18,7 @@
             <li id="notification"><a href="#"></a></li>
         </ul>
     </header>
-    <form class="login_form" id="loginForm" method="GET" action="/">
+    <form class="login_form" id="loginForm" method="POST" action="/">
```
Comment (8o8inCodes) [inline public/login/login.html:21]: Avoid using form functionality, create functions in js that would handle the submit button. and make sure that you add "preventsDefault()" so that it won't do form's functionality.

File: public/register/register.html [hunk 15]
```diff
@@ -17,17 +17,9 @@
             <li><a href="#" class="active">Register</a></li>
         </ul>
     </header>
-    <form class="register_form" id="register_Form" method="get" action="/">
+    <form class="register_form" method="POST" action="/register">
```
Comment (8o8inCodes) [inline public/register/register.html:20]: Same here

File: routes/register.js [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+        username,
+        password: hashedPass,
+        email,
+        lastName,
+        firstName
+    })
+    await user.save();
+    res.redirect('/')
+});
+    module.exports = router;
```
Comment (8o8inCodes) [inline routes/register.js]: Too many duplicates, you already have 2 register routes and another as a controller, what on earth is going on lol

**Review by 8o8inCodes — CHANGES_REQUESTED** [review]

File: api/registerRouter.js (hunk 1 above)
Comment (Nikitak211) [reply api/registerRouter.js]: its deleted at posterJS  branch

File: app.js (hunk 5 above)
Comment (Nikitak211) [reply app.js]: fixed at publicJS

File: routes/register.js (hunk 16 above)
Comment (Nikitak211) [reply routes/register.js]: its the one that will work on publicJS

File: public/login/login.html (hunk 14 above)
Comment (Nikitak211) [reply public/login/login.html:21]: will fix in the next branch.

File: public/register/register.html (hunk 15 above)
Comment (Nikitak211) [reply public/register/register.html:20]: next branch will resolve

File: .env-sample [hunk 17]
```diff
@@ -0,0 +1,2 @@
+SecretToken = write somthing that is difficult to hack.
+MongoDBUri = enter your data uri for mongodb.
```
Comment (8o8inCodes) [inline .env-sample]: Environment variables usually all caps. and give better name to secret token. You will have more tokens in the future.
JWT_SECRET
Would be better

File: app.js [hunk 18]
```diff
@@ -1,25 +1,70 @@
...
 
-//body parser Midware
-app.use(express.json());
-app.use(express.urlencoded({ extended: false }));
+// .env confing
+require('dotenv').config();
 
-app.use('/static', express.static(path.join(__dirname, 'public')))
+//mongoose promise
+mongoose.Promise = global.Promise
```
Comment (8o8inCodes) [inline app.js]: What is this?

File: app.js [hunk 19]
```diff
@@ -1,25 +1,70 @@
...
+    console.log('***connection not available***', error);
+});
+
+const store = new MongoDBStore({
+    uri: mongoDBUri,
+    collection: 'database'
+});
+
+app.use(session({
+    secret: 'slob that bob top',
```
Comment (8o8inCodes) [inline app.js]: This also has to be in env. SESSION_SECRET

File: package.json [hunk 20]
```diff
@@ -17,11 +17,15 @@
...
+    "dotenv": "^10.0.0",
     "express": "^4.17.1",
+    "express-session": "^1.17.2",
+    "jsonwebtoken": "^8.5.1",
     "mongodb": "^4.1.2",
     "mongoose": "^6.0.7",
     "node": "^16.9.1",
-    "nodemon": "^2.0.13",
-    "uuid": "^8.3.2"
+    "nodemon": "^2.0.13"
```
Comment (8o8inCodes) [inline package.json:30]: Nodemon should be in dev dependencies

**Review by 8o8inCodes — CHANGES_REQUESTED** [review]

File: .env-sample [hunk 21]
```diff
@@ -0,0 +1,3 @@
+JWT_SECRET = write somthing that is difficult to hack.
+MONGODB_URI = enter your data uri for mongodb.
+SECRET = a secret for the session use. write somthing that is difficult to hack.
```
Comment (8o8inCodes) [inline .env-sample:3]: Don't call it just secret, you need to specify which secret this is.
SESSION_SECRET

File: app.js [hunk 22]
```diff
@@ -1,25 +1,67 @@
+//Third parte packages
```
Comment (8o8inCodes) [inline app.js]: party*

File: app.js [hunk 23]
```diff
@@ -1,25 +1,67 @@
...
+const MongoDBStore = require('connect-mongodb-session')(session);
 
-//body parser Midware
-app.use(express.json());
-app.use(express.urlencoded({ extended: false }));
+// .env confing
+require('dotenv').config();
 
-app.use('/static', express.static(path.join(__dirname, 'public')))
+//My imports
```
Comment (8o8inCodes) [inline app.js]: Call it "Local imports"

File: package.json [hunk 24]
```diff
@@ -17,11 +17,17 @@
...
+    "dotenv": "^10.0.0",
     "express": "^4.17.1",
+    "express-session": "^1.17.2",
+    "jsonwebtoken": "^8.5.1",
     "mongodb": "^4.1.2",
     "mongoose": "^6.0.7",
-    "node": "^16.9.1",
-    "nodemon": "^2.0.13",
-    "uuid": "^8.3.2"
+    "node": "^16.9.1"
```
Comment (8o8inCodes) [inline package.json]: Why did you add node here? it's not required lol

**Review by 8o8inCodes — CHANGES_REQUESTED** [review]

File: package.json (hunk 24 above)
Comment (Nikitak211) [reply package.json]: honestly dont know lol used to
